### PR TITLE
TrecRun.remove/keep docs by qrels

### PR DESCRIPTION
**Pull request by x65han**
# Add two functions to TrecRun

- keep docs by qrels so that we get 100% judged
- remove docs by qrels so e.g. we don't want rnd1 docs when submitting round 2

### CODE REVIEW

**Comment by lintool** [comment]
hey @x65han is this meant to be a replacement for these scripts?

**Comment by x65han** [comment]
No replacement is happening. I am moving the code around a little bit.


1. I moved Qrels to be before TrecRun, but still in the same file. This way, we can refer to the Qrels type in TrecRun.

2. I also moved a couple test cases from test_fusion.py to test_trectools.py since the two test cases are testing Qrels & TrecRun class, which belong to the trectools submodules.

**Comment by lintool** [comment]
Sorry, it appears that I missed the URL: https://github.com/castorini/anserini-tools/blob/master/scripts/filter_run_with_qrels.py

It seems like the filter qrels portion is a cleaner, more integration version of the above script?

File: pyserini/trectools/_base.py [hunk 1]
```diff
@@ -99,6 +152,71 @@ def rescore(self, method: RescoreMethod, rrf_k: int = None, scale: float = None)
     def to_numpy(self) -> np.ndarray:
         return self.run_data.to_numpy(copy=True)
 
+    def remove_by_qrels(self, qrels: Qrels, copy=True):
```
Comment (lintool) [inline pyserini/trectools/_base.py:155]: how about `discard_qrels` and `retain_qrels` to be consistent with https://github.com/castorini/anserini-tools/blob/master/scripts/filter_run_with_qrels.py ?

File: pyserini/trectools/_base.py [hunk 2]
```diff
@@ -99,6 +152,71 @@ def rescore(self, method: RescoreMethod, rrf_k: int = None, scale: float = None)
...
+        """Given a Qrels object, remove each docid in self if docid is also in the qrels.
+        This operation is performed on each topic separately.
```
Comment (lintool) [inline pyserini/trectools/_base.py:157]: I think docstring convention is to begin description with a verb, so something like "Discard docids that are present in a Qrels object..."

File: pyserini/trectools/_base.py [hunk 3]
```diff
@@ -99,6 +152,71 @@ def rescore(self, method: RescoreMethod, rrf_k: int = None, scale: float = None)
...
+
+        Parameters:
+        ----------
+        qrels : Qrels
+            Qrels with docids to remove from TrecRun.
+        copy : Bool
```
Comment (lintool) [inline pyserini/trectools/_base.py:163]: I think `clone` is more standard?

File: pyserini/trectools/_base.py [hunk 4]
```diff
@@ -99,6 +152,71 @@ def rescore(self, method: RescoreMethod, rrf_k: int = None, scale: float = None)
...
+
+    def _filter_from_qrels(self, qrels: Qrels, keep: bool, copy=True):
+        """Private helper function to remove/keep each docid in self if docid is also in the given Qrels object.
+        This operation is performed on each topic separately.
+
+        Parameters:
+        ----------
+        qrels : Qrels
+            Qrels with docids to remove from or keep in TrecRun.
+        copy : Bool
```
Comment (lintool) [inline pyserini/trectools/_base.py:192]: typo here? two "copy"

File: pyserini/trectools/_base.py [hunk 5]
```diff
@@ -99,6 +152,71 @@ def rescore(self, method: RescoreMethod, rrf_k: int = None, scale: float = None)
...
+            Return a new TrecRun object if True, else self will be modified and returned.
+        """
+
+        df_list = []
+        for topic in self.topics():
+            if topic not in qrels.topics():
+                continue
+
+            tabu_list = qrels.get_docids(topic)
```
Comment (lintool) [inline pyserini/trectools/_base.py:203]: what's "tabu" mean?

Comment (x65han) [reply pyserini/trectools/_base.py:203]: Coming from [Tabu Search](https://www.sciencedirect.com/topics/computer-science/tabu-search#:~:text=2.2%20Tabu%20Search%20Algorithm%20(TS,composed%20by%20the%20next%20elements.), tabu just means something we don't want. Maybe not the best naming here. I am going to name it `qrels_docids`.

**Comment by x65han** [comment]
> Sorry, it appears that I missed the URL: https://github.com/castorini/anserini-tools/blob/master/scripts/filter_run_with_qrels.py
> 
> It seems like the filter qrels portion is a cleaner, more integration version of the above script?

Yes, integrating the logic of `filter_run_with_qrels.py` into our TrecRun and Qrels object.

**Pull request merged**